File: Python/Search_a_2D_Matrix.py
"""
Write an efficient algorithm that searches for a value in an m x n matrix. This matrix has the following properties:

Integers in each row are sorted from left to right.
The first integer of each row is greater than the last integer of the previous row.
"""
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix:
            return 0
        s=0
        e=(len(matrix)*len(matrix[0]))-1
        while(s<=e):
            mid=(s+e)//2 
            r=mid//len(matrix[0])
            c=mid%len(matrix[0])
            if matrix[r][c]==target:
                return 1
            elif matrix[r][c]>target:
                e=mid-1
            else:
                s=mid+1
        return 0

File: folders/LeetCode C++ Solutions/search-a-2d-matrix.cpp
/**
 * Author: Siddhant Pandya
 * Problem: 74. Search a 2D Matrix
 */

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(!matrix.empty() && !matrix[0].empty())
        {
        for(int x=0;x<matrix.size();x++)
        {
            if(matrix[x][matrix[x].size()-1]>=target)
            {
                for(int y=0;y<matrix[x].size();y++)
                {
                    if(matrix[x][y]==target)
                    {
                        return true;
                    }
                }
                break;
            }
        }
        }
        
        return false;
    }
};